src/lib.rs

```diff
@@ -0,0 +1,560 @@
+            self.nodes[index].left_child = self.nodes[left_index].right_child;
+            if let Some(new_left_index) = self.nodes[index].left_child {
+                self.nodes[new_left_index].parent = Some(index);
+            }
+            self.nodes[left_index].right_child = Some(index);
+            let parent = self.nodes[index].parent;
+            if let Some(parent_index) = parent {
+                if self.nodes[parent_index].left_child == Some(index) {
+                    self.nodes[parent_index].left_child = Some(left_index);
+                } else {
+                    self.nodes[parent_index].right_child = Some(left_index);
+                }
+            } else {
+                // change the root
+                self.root = Some(left_index);
+            }
+            self.nodes[index].height =
+                self.compute_height(self.nodes[index].left_child,
+                                    self.nodes[index].right_child);
+            self.nodes[left_index].height =
+                self.compute_height(self.nodes[left_index].left_child,
+                                    self.nodes[left_index].right_child);
+            self.nodes[left_index].parent = self.nodes[index].parent;
+            self.nodes[index].parent = Some(left_index);
+            self.update_heights(index);
+        }
+    }
+
+    fn right_left_rotation(&mut self, index: usize) {
+        if let Some(right_index) = self.nodes[index].right_child {
+            self.right_rotation(right_index);
+            self.left_rotation(index);
+        }
+    }
+
+    fn left_right_rotation(&mut self, index: usize) {
+        if let Some(left_index) = self.nodes[index].left_child {
+            self.left_rotation(left_index);
+            self.right_rotation(index);
+        }
+    }
+
+    fn rebalance(&mut self, index: usize) {
+        let balance_factor = self.balance_factor(&self.nodes[index]);
+        if balance_factor == -2 {
+            // right-heavy tree
+            let right_index = self.nodes[index].right_child.unwrap();
+            if self.balance_factor(&self.nodes[right_index]) == -1 {
+                self.left_rotation(index);
+            } else {
+                self.right_left_rotation(index);
+            }
+        } else if balance_factor == 2 {
+            // left-heavy
+            let left_index = self.nodes[index].left_child.unwrap();
+            if self.balance_factor(&self.nodes[left_index]) == 1 {
+                self.right_rotation(index);
+            } else {
+                self.left_right_rotation(index);
+            }
+        } else if balance_factor < -2 || balance_factor > 2 {
+            panic!("Invalid balance factor for {}: {}", index, balance_factor);
+        }
+    }
+
+    fn is_balanced(&self, index: usize) -> bool {
+        let balance_factor = self.balance_factor(&self.nodes[index]);
+        balance_factor >= -1 && balance_factor <= 1
```

Comment:
Can you chain operators in Rust?

```
return -1 <= self.balance_factor(&self.nodes[index]) <= 1
```

Reply:
If you can't you could do:

```
return abs(self.balance_factor(&self.nodes[index])) < 2
```

Reply:
It seems to be an [open issue](https://github.com/rust-lang/rfcs/issues/2083), I'll just use `abs`